Implement C++ API version of torch.nn.functional.one_hot (#27081)

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27177 Implement C++ API version of torch.nn.functional.one_hot (#27081)**

Summary:
Add support for F::one_hot C++ function.

Test Plan:
Added 3 new tests to verify API is working

Reviewers: ZolotukhinM

Differential Revision: [D17697934](https://our.internmc.facebook.com/intern/diff/D17697934)

bzinodev committed:
Implement C++ API version of torch.nn.functional.one_hot (#27081)

Summary:
Add support for F::one_hot C++ function.

Test Plan:
Added 3 new tests to verify API is working

Reviewers: ZolotukhinM

[ghstack-poisoned]

bzinodev committed:
Update on "Implement C++ API version of torch.nn.functional.one_hot (#27081)"


Summary:
Add support for F::one_hot C++ function.

Test Plan:
Added 3 new tests to verify API is working

Reviewers: ZolotukhinM

Differential Revision: [D17697934](https://our.internmc.facebook.com/intern/diff/D17697934)

[ghstack-poisoned]

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -184,3 +186,42 @@ TEST_F(FunctionalTest, MaxUnpool1d) {
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_TRUE(torch::allclose(y, expected));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({3, 2, 3}));
+}
```
nit: we can put all of these three tests into one test case:
```cpp
TEST_F(FunctionalTest, OneHot) {
  {
    // Code in OneHot_1
  }
  {
    // Code in OneHot_2
  }
  {
    // Code in OneHot_3
  }
}
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/embedding.h:
```diff
@@ -0,0 +1,12 @@
+namespace nn {
+namespace functional {
+
+inline Tensor one_hot(const Tensor& x, int64_t num_classes = -1) {
```
nit: naming it `tensor` instead of `x` would be awesome, to match the Python version :)

yf225 reviewed (commented):
@bzinodev Thanks so much for the great work! It looks awesome and I left some very minor comments.

bzinodev committed:
Update on "Implement C++ API version of torch.nn.functional.one_hot (#27081)"


Summary:
Add support for F::one_hot C++ function.

Test Plan:
Added 3 new tests to verify API is working

Reviewers: ZolotukhinM

Differential Revision: [D17697934](https://our.internmc.facebook.com/intern/diff/D17697934)

[ghstack-poisoned]

bzinodev committed:
Update on "Implement C++ API version of torch.nn.functional.one_hot (#27081)"


Summary:
Add support for F::one_hot C++ function.

Test Plan:
Added 3 new tests to verify API is working

Reviewers: ZolotukhinM

Differential Revision: [D17697934](https://our.internmc.facebook.com/intern/diff/D17697934)

[ghstack-poisoned]

yf225 reviewed (approved):
Looks awesome! Thanks a lot @bzinodev :D
